research/fed-sm/job_configs/fedsm_prostate/app/custom/helpers/pt_fedsm.py

```diff
@@ -0,0 +1,195 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from abc import abstractmethod
+
+import numpy as np
+import torch
+
+from nvflare.apis.fl_constant import ReturnCode
+from nvflare.apis.shareable import make_reply
+from nvflare.apis.signal import Signal
+
+
+def AccuracyTopK(output, target, topk=(1,)):
+    """Computes the accuracy over the k top predictions for the specified values of k"""
+    with torch.no_grad():
+        maxk = max(topk)
+        batch_size = target.size(0)
+
+        _, pred = output.topk(maxk, 1, True, True)
+        pred = pred.t()
+        correct = pred.eq(target.view(1, -1).expand_as(pred))
+
+        res = []
+        for k in topk:
+            correct_k = correct[:k].reshape(-1).float().sum(0, keepdim=True)
+            res.append(correct_k.mul_(100.0 / batch_size))
+        return res
+
+
+class PTFedSMHelper(object):
+    def __init__(
+        self,
+        person_model,
+        select_model,
+        person_criterion,
+        select_criterion,
+        person_optimizer,
+        select_optimizer,
+        device,
+        app_dir,
+        person_model_epochs: int = 1,
+        select_model_epochs: int = 1,
+    ):
+        """Helper to be used with FedSM components.
+        Implements the functions used for the algorithm proposed in
+        Xu et al. "Closing the Generalization Gap of Cross-silo Federated Medical Image Segmentation"
+        (https://arxiv.org/abs/2203.10144) using PyTorch.
+
+        Args:
+            person/select_model: the personalized and selector models
+            person/select_criterion: loss criterion
+            person/select_optimizer: training optimizer the two models
+            device: device for model training
+            app_dir: needed for local personalized model saving
+            person/select_model_epochs: total training epochs each round
+
+        Returns:
+            None
+        """
+
+        self.person_model = person_model
+        self.select_model = select_model
+        self.person_criterion = person_criterion
+        self.select_criterion = select_criterion
+        self.person_optimizer = person_optimizer
+        self.select_optimizer = select_optimizer
+        if device is None:
+            self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        else:
+            self.device = device
+        self.person_model_epochs = person_model_epochs
+        self.select_model_epochs = select_model_epochs
+        # check criterion, model, and optimizer type
+        if not isinstance(self.person_model, torch.nn.Module):
+            raise ValueError(f"person_model component must be torch model. " f"But got: {type(self.person_model)}")
```

Comment:
The strings in these ValueErrors are formatted incorrectly.

Reply:
thanks! corrected